cpp/src/ice2slice/Gen.h

```diff
@@ -22,35 +22,53 @@ namespace Slice
-            virtual bool visitStructStart(const StructPtr&);
-            virtual void visitSequence(const SequencePtr&);
-            virtual void visitDictionary(const DictionaryPtr&);
-            virtual void visitEnum(const EnumPtr&);
+            bool visitClassDefStart(const ClassDefPtr&) final;
+            bool visitInterfaceDefStart(const InterfaceDefPtr&) final;
+            bool visitExceptionStart(const ExceptionPtr&) final;
+            bool visitStructStart(const StructPtr&) final;
+            void visitSequence(const SequencePtr&) final;
+            void visitDictionary(const DictionaryPtr&) final;
+            void visitEnum(const EnumPtr&) final;
 
+            /// After visiting all definitions, return the list of scopes corresponding to modules for which we need
+            /// to generate an output `.slice`.
+            /// @return The list of modules for which we need to generate an output `.slice`.
             std::set<std::string> modules() const;
 
         private:
             std::set<std::string> _modules;
         };
 
-        class TypesVisitor : public ParserVisitor
+        /// The TypesVisitor class converts Slice definitions in `.ice` files to corresponding Slice definitions in
+        /// `.slice` files.
+        ///
+        /// If the `.ice` file contains a single module, the output is a single `.slice` file with the same base name,
+        /// replacing the `.ice` extension with `.slice`.
+        ///
+        /// If the `.ice` file contains multiple modules, it generates a `.slice` file for each module. Each file uses
+        /// the base name of the `.ice` file, appending the module scope with "::" replaced by "_" and the `.slice`
+        /// extension.
+        class TypesVisitor final : public ParserVisitor
         {
         public:
-            TypesVisitor(const std::string&, const std::set<std::string>&);
+            TypesVisitor(const std::string& fileBase, const std::set<std::string>& modules);
 
-            virtual bool visitClassDefStart(const ClassDefPtr&);
-            virtual bool visitInterfaceDefStart(const InterfaceDefPtr&);
-            virtual bool visitExceptionStart(const ExceptionPtr&);
-            virtual bool visitStructStart(const StructPtr&);
-            virtual void visitSequence(const SequencePtr&);
-            virtual void visitDictionary(const DictionaryPtr&);
-            virtual void visitEnum(const EnumPtr&);
+            bool visitClassDefStart(const ClassDefPtr&) final;
+            bool visitInterfaceDefStart(const InterfaceDefPtr&) final;
+            bool visitExceptionStart(const ExceptionPtr&) final;
+            bool visitStructStart(const StructPtr&) final;
+            void visitSequence(const SequencePtr&) final;
+            void visitDictionary(const DictionaryPtr&) final;
+            void visitEnum(const EnumPtr&) final;
+            void visitConst(const ConstPtr&) final;
+
+            void newLine(); // Ensure all files end with a newline
```

Comment:
I would move this code/logic to `visitUnitEnd` - the end of the compilation unit.